### PR TITLE
Zacnotmaster

### PR DESCRIPTION
**Pull request by ZacLN**
Pulls in hover docs, completions, function signature help and moves the linter and file operations to the Request-Respond model

### CODE REVIEW

File: scripts/languageserver/languageserver.jl [hunk 1]
```diff
@@ -6,39 +6,40 @@ if VERSION < v"0.5"
 end
 
 include("dependencies.jl")
-use_and_install_dependencies(["Compat", "JSON", "Lint", "URIParser"])
+#use_and_install_dependencies(["Compat", "JSON", "Lint", "URIParser","JuliaParser"])
+using Compat,JSON,Lint,URIParser,JuliaParser
```
Comment (davidanthoff) [inline scripts/languageserver/languageserver.jl]: Can you use ``use_and_install_dependencies`` for everything here?

File: scripts/languageserver/misc.jl [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return Response{didOpen,Exception}("2.0",r.id,err)
+    end
+end
+
+abstract didClose <: Method
+
+function Respond(r::Request{didClose,TextDocumentIdentifier})
+    try
+        delete!(documents,r.params.uri)
+        info(r.params.uri)
```
Comment (davidanthoff) [inline scripts/languageserver/misc.jl]: This can go?

File: scripts/languageserver/protocol.jl [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+#    finish::Position
+#end
+let ex=:(type Range
+        start::Position
+        finish::Position
+    end)
+    ex.args[3].args=ex.args[3].args[[2;4]]
+    ex.args[3].args[2].args[1]=Symbol("end")
+    eval(ex)
+end
```
Comment (davidanthoff) [inline scripts/languageserver/protocol.jl]: This seems very complicated, is that necessary?

File: scripts/languageserver/protocol.jl [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+getSym(p::TextDocumentPositionParams) = getSym(Word(p))
+
+
+function checkmsg(response)
+    n = length(response)
+    io = IOBuffer()
+    write(io, "Content-Length: $n\r\n\r\n")
+    write(io, response)
+    takebuf_string(io)
+end
```
Comment (davidanthoff) [inline scripts/languageserver/protocol.jl]: This is never used, can this go?

File: scripts/languageserver/protocol.jl [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+function first20lines(d::String)
+    cnt = 0
+    l10 = 0
+    @inbounds @simd for i =1:length(d.data)
+        cnt==20 && (l10=i)
+        cnt+=d.data[i]==0x0a
+    end
+    return cnt,1:l10-1
+end
```
Comment (davidanthoff) [inline scripts/languageserver/protocol.jl]: Same here, seems never used, can this go?

File: scripts/languageserver/transport.jl [hunk 6]
```diff
@@ -18,7 +18,8 @@ function read_transport_layer(stream)
 end
 
 function write_transport_layer(stream, response)
-    n = length(response)
+    response_utf8 = transcode(UInt8, response)
+    n = length(response_utf8)
     write(stream, "Content-Length: $n\r\n\r\n")
-    write(stream, response)
+    write(stream, response_utf8)
```
Comment (davidanthoff) [inline scripts/languageserver/transport.jl]: I'm confused what is going on here in git terms. I thought I had committed this to master, why is this showing up here again? Maybe you need to rebase your branch against master? Not sure...

File: src/extension.ts [hunk 7]
```diff
@@ -61,12 +61,13 @@ function loadConfiguration() {
 function startLanguageServer(context: vscode.ExtensionContext) {
     // let debugOptions = { execArgv: ["--nolazy", "--debug=6004"] };
 
-    var originalJuliaPkgDir = process.env.JULIA_PKGDIR ? process.env.JULIA_PKGDIR : path.join(os.homedir(), '.julia', 'v0.5');
+    var originalJuliaPkgDir = process.env.JULIA_PKGDIR ? process.env.JULIA_PKGDIR : path.join(os.homedir(), '.julia', 'v0.6');
```
Comment (davidanthoff) [inline src/extension.ts]: This should stay 0.5 for now, I guess. I think once we have merged this PR we can do better here to detect the correct folder, I think I know how that should be done. But for now, lets just keep this at 0.5 and then merge this, ok?

File: src/extension.ts [hunk 8]
```diff
@@ -61,12 +61,13 @@ function loadConfiguration() {
...
     let serverArgs = ['--startup-file=no', '--history-file=no', 'languageserver.jl', originalJuliaPkgDir];
     let spawnOptions = {
         cwd: path.join(context.extensionPath, 'scripts', 'languageserver'),
         env: {
-            JULIA_PKGDIR: path.join(context.extensionPath, 'scripts', 'languageserver', 'julia_pkgdir'),
+            JULIA_PKGDIR: "/home/zac/.julia/v0.6",
+            // JULIA_PKGDIR: path.join(context.extensionPath, 'scripts', 'languageserver', 'julia_pkgdir'),
```
Comment (davidanthoff) [inline src/extension.ts]: These edits should all not be part of this PR.

File: scripts/languageserver/languageserver.jl (hunk 1 above)
Comment (ZacLN) [inline scripts/languageserver/languageserver.jl]: Sorry, I didn't mean to merge that and did mean to tell you that I can't get this function to work, I'll have a look to see why later.

File: scripts/languageserver/protocol.jl (hunk 3 above)
Comment (ZacLN) [inline scripts/languageserver/protocol.jl]: Extremely ugly, I just bodged it together. Do you know of a way to use a keyword as a fieldname?

Comment (ZacLN) [inline scripts/languageserver/protocol.jl]: for ref [#2665](https://github.com/JuliaLang/julia/issues/2665)

File: scripts/languageserver/transport.jl (hunk 6 above)
Comment (ZacLN) [inline scripts/languageserver/transport.jl]: Ok i _think_ I'm on level terms now? If not let me know and I'll re-fork the upstream
